### PR TITLE
Add parsing methods to Range aggregations

### PR DESCRIPTION
**Pull request by tlrx**
This pull request adds the parsing logic to InternalRange, InternalDateRange and InternalGeoDistance aggregations.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/range/ParsedRange.java [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+                builder.startObject();
+                builder.field(CommonFields.KEY.getPreferredName(), key);
+            }
+            if (!Double.isInfinite(from)) {
+                builder.field(CommonFields.FROM.getPreferredName(), from);
+                if (fromAsString != null) {
+                    builder.field(CommonFields.FROM_AS_STRING.getPreferredName(), fromAsString);
+                }
+            }
+            if (!Double.isInfinite(to)) {
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/range/ParsedRange.java]: nit: == false?

**Review by javanna — APPROVED** [review]
LGTM

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/range/ParsedRange.java [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    public static ParsedRange fromXContent(XContentParser parser, String name) throws IOException {
+        ParsedRange aggregation = PARSER.parse(parser, null);
+        aggregation.setName(name);
+        return aggregation;
+    }
+
+    public static class ParsedBucket extends ParsedMultiBucketAggregation.ParsedBucket implements Range.Bucket {
+
+        protected String key;
+        protected double from;
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/range/ParsedRange.java]: This should probably be initialized to Double.NEGATIVE_INFINITY. If `InternalRange.from` is infinite, the "from" field won't be rendered in the xContent and I think the current parser would leave from == 0 in this case. Maybe we should also add tests for those edge cases.

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/range/ParsedRange.java [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+        protected String fromAsString;
+        protected double to;
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/range/ParsedRange.java]: Same here but with Double.POSITIVE_INFINITY?

**Review by cbuescher — CHANGES_REQUESTED** [review]
Looks good, but I found two edge cases around the range to/from value that might not be covered by tests yet and that might not be working as expected yet. Not entirely sure, but maybe worth taking another close look.

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/range/ParsedRange.java (hunk 2 above)
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/range/ParsedRange.java]: That's right, thanks! I added some random ranges with infinity values to the test.

**Comment by tlrx** [comment]
@cbuescher I updated the code. Can you please have another look? Thanks

**Review by cbuescher — APPROVED** [review]
LGTM

**Comment by tlrx** [comment]
Thanks @cbuescher @javanna
